nova/virt/vmwareapi/vmops.py

```diff
@@ -2421,21 +2640,164 @@ def live_migration(self, instance, migrate_data, volume_mapping):
         vm_util.relocate_vm(self._session, vm_ref, spec=relocate_spec)
 
     def _detach_volumes(self, instance, block_device_info):
-        block_devices = driver.block_device_info_get_mapping(block_device_info)
-        for disk in block_devices:
-            self._volumeops.detach_volume(disk['connection_info'], instance)
-
-    def _attach_volumes(self, instance, block_device_info, adapter_type):
         disks = driver.block_device_info_get_mapping(block_device_info)
-        # make sure the disks are attached by the boot_index order (if any)
+        # Detach the volumes in reverse order, so if we roll it back
+        # that the device order will still be preserved
+        for disk in sorted(disks,
+                           reverse=True,
+                           key=itemgetter('mount_device')):
```

Comment:
Yes,  I am not too happy about the field, but I think it is the most reliable way to ensure a deterministic order of the devices.
The `boot_index` used before is rarely set.

Reply:
The `id` of the BDM in the DB could help or the `created_at` - but I'm not sure any of those make it into the object we have here.

Reply:
Why I relied on the device-name:

- First disk is sda
- Second disk is sdb
- Third disk is sdc
- Removing the second disk
- Add another third disk, and it will be sdb

If I put them in the list in the order of id or creation,
the disks sdc and sdb will be swapped.

Why I think, the whole device naming in the API doesn't make sense:

Reboot after removing the second disk, and you end up with sda and sdb,
Add now a third disk, it will appear as sdc. And if we use the LUN 1, that will flip again after reboot.

Reply:
Fair point. Didn't think of that.

Agreed, that the whole device naming doesn't make sense. I think I read somewhere upstream, that since no driver able to use this information, they want to remove it in the future. Something, we need a solution for then (and not now :D).